kernel/kernel-api/src/main/java/io/delta/kernel/internal/checksum/ChecksumUtils.java

```diff
@@ -169,6 +201,149 @@ private static CRCInfo buildCrcInfoWithFullLogReplay(
         Optional.of(state.addedFileSizeHistogram));
   }
 
+  /**
+   * Attempts to build CRC info incrementally from the last seen checksum. Falls back if incremental
+   * computation is not possible.
+   *
+   * @param lastSeenCrcInfo The last available CRC info to build upon
+   * @param engine The engine to use for file operations
+   * @param logSegment The log segment to process
+   * @return Optional containing the new CRC info, or empty if fallback is needed
+   */
+  private static Optional<CRCInfo> buildCrcInfoIncrementally(
+      CRCInfo lastSeenCrcInfo, Engine engine, LogSegment logSegment) throws IOException {
+    // Can only build incrementally if we have domain metadata and file size histogram
+    if (!lastSeenCrcInfo.getDomainMetadata().isPresent()) {
+      logger.info("Falling back to full replay: detected current crc missing domain metadata.");
+      return Optional.empty();
+    }
+    if (!lastSeenCrcInfo.getFileSizeHistogram().isPresent()) {
+      logger.info("Falling back to full replay: detected current crc missing file size histogram.");
+      return Optional.empty();
+    }
+
+    // Initialize state tracking
+    StateTracker state = new StateTracker();
+
+    // Create iterator for delta files newer than last CRC
+    try (CloseableIterator<ColumnarBatch> iterator =
+        DeltaLogActionUtils.readCommitFiles(
+            engine,
+            logSegment.getDeltas().stream()
+                .filter(
+                    file ->
+                        FileNames.getFileVersion(new Path(file.getPath()))
+                            > lastSeenCrcInfo.getVersion())
+                .sorted(
+                    Comparator.comparing((FileStatus a) -> new Path(a.getPath()).getName())
+                        .reversed())
+                .collect(Collectors.toList()),
+            CHECKPOINT_SCHEMA.add("commitInfo", CommitInfo.FULL_SCHEMA))) {
+
+      Optional<Long> currentVersion = Optional.empty();
+
+      while (iterator.hasNext()) {
+        ColumnarBatch batch = iterator.next();
+        final int rowCount = batch.getSize();
+        if (rowCount == 0) {
+          continue;
+        }
+
+        StructType schema = batch.getSchema();
+        ColumnVector versionVector = batch.getColumnVector(schema.indexOf("version"));
+        ColumnVector addVector = batch.getColumnVector(schema.indexOf("add"));
+        ColumnVector removeVector = batch.getColumnVector(schema.indexOf("remove"));
+        ColumnVector metadataVector = batch.getColumnVector(schema.indexOf("metaData"));
+        ColumnVector protocolVector = batch.getColumnVector(schema.indexOf("protocol"));
+        ColumnVector domainMetadataVector = batch.getColumnVector(schema.indexOf("domainMetadata"));
+        ColumnVector commitInfoVector = batch.getColumnVector(schema.indexOf("commitInfo"));
```

Comment:
schema.indexOf can be above the while loop. Actually declare both the schema and indices as static members.